utils/s3/client.cc

```diff
@@ -854,6 +858,167 @@ data_sink client::make_upload_jumbo_sink(sstring object_name, std::optional<unsi
     return data_sink(std::make_unique<upload_jumbo_sink>(shared_from_this(), std::move(object_name), max_parts_per_piece));
 }
 
+// a sink with known object size
+//
+// unlike upload_sink and upload_jumbo_sink, upload_ranged_sink sends the
+// received buffers right away without accumulating them first. if the
+// buffer fed by the caller crosses the boundary of the current part, a
+// new part is created, and the remainder of the buffer is sent to it instead,
+// until whole buffer is consumed.
+class client::upload_ranged_sink final : public upload_sink_base {
+    // TODO: resume support
+    const size_t _part_size;
+    const size_t _total_size;
+    std::optional<future<>> _uploading;
+
+    // avoid stuffing the memory with yet-uploaded buffers in the expense of
+    // having ping-pong between reader (from disk) and writer (to the wire)
+    seastar::pipe<temporary_buffer<char>> _part_body{16};
```

Comment:
This might need some more care. There's `client::_memory` semaphore that controls the amount of memory sinks can hold in RAM. This creates independent limit which is per-sink, not per-client, so you can have 1000 sinks thus eating more memory than expected

Reply:
s/16/1/ so we don't hold the memory chunks here.